### PR TITLE
[bug-fix] enter `tf.init_scope` when adding optimizer variables with TF backend

### LINKED ISSUE #23085: [Bug] TF tracing fails for distributed training with stateful optimizers
Hi!

When trying to use a "stateful" optimizer (i.e. with optimizer variables, e.g. SGD with momentum, Adam) within a custom TF loop in distributed training, I keep getting these TF graph/tracing errors:

```
2 root error(s) found.
  (0) INVALID_ARGUMENT:  You must feed a value for placeholder tensor 'cond/else/_1/cond/Placeholder' with dtype int32
         [[{{node cond/Placeholder}}]]
         [[Func/cond/then/_0/input/_29/_16]]
  (1) INVALID_ARGUMENT:  You must feed a value for placeholder tensor 'cond/else/_1/cond/Placeholder' with dtype int32
         [[{{node cond/Placeholder}}]]
```

This is the same error as in
- #19246 which seems to have been solved since from the discussions, but used the built-in training loop with stateless optimizers
- #20164 which used a custom training loop and a stateful optimizer and still fails, and for which I found that using a stateless optimizer or applying my proposed fix below works

## Reproduction
On latest master or v3.14.1 + TF 2.21.0, but I've seen it with older versions as well.
A logical device distributed context is enough to reproduce, no need for actual GPUs.
A very simple model is also enough.
```py
import keras
import tensorflow as tf

tf.config.set_logical_device_configuration(
    tf.config.list_physical_devices('CPU')[0],
    [tf.config.LogicalDeviceConfiguration()]*2
)

class Model(keras.Model):
    def __init__(self):
        super().__init__()
        self.w = self.add_weight()
    def call(self, x):
        return self.w * x

# cross_device_ops=tf.distribute.ReductionToOneDevice() works
strategy = tf.distribute.MirroredStrategy()
with strategy.scope():
    inputs, targets = tf.random.normal((1,)), tf.random.normal((1,))
    opt = keras.optimizers.SGD(momentum=.1) # momentum=0. works
    model = Model()

@tf.function # removing decorator works
def replica_fn(inputs, targets):
    with tf.GradientTape() as tape:
        pred = model(inputs)
        loss = keras.losses.mean_absolute_error(pred, targets)
    grads = tape.gradient(loss, model.trainable_variables)
    opt.apply_gradients(zip(grads, model.trainable_variables))

strategy.run(replica_fn, args=(inputs, targets))

```

## Workarounds
- Set `cross_device_ops=tf.distribute.ReductionToOneDevice()` in `MirroredStrategy`
- Use state/variable-less optimizers, e.g. SGD with no momentum
- Use a single device
- Remove `@tf.function`

## Suggested fix
While going through the optimizers / TF backend code with the help of an AI assistant, it proposed the use of [`tf.init_scope`](https://www.tensorflow.org/api_docs/python/tf/init_scope) when creating optimizer variables.
Wrapping `add_variable` within a `tf.init_scope` context in `TFOptimizer` indeed solves my issue and seems to keep optimizer tests OK.
Something like:
```py
def add_variable(self, *args, **kwargs):
    with tf.init_scope():
        return super().add_variable(*args, **kwargs)

```

My understanding is that `tf.init_scope` is made for this type of issue 
> There is often a need to lift variable initialization ops out of control-flow scopes, function-building graphs, and gradient tapes.

but I must admit I'm not 100% familiar with it to see if there could be any unintended side effects.

If you think this solution is sound, I'd be happy to open a PR for it.
Otherwise let me know if there's anything I can do to help.

Thank you

### PR DESCRIPTION
**Pull request by peinguim**
## Description

As discussed in #23085 (thank you @maitry63 for taking the time to respond), this fixes a tracing error in TF distributed training with stateful optimizers by entering a `tf.init_scope()` when creating optimizer variables.

Fixes #23085.

AI use: to help go through the optimizers code + it proposed the use of `tf.init_scope()`

## Contributor Agreement

**Please review our [AI-Assisted Contribution Policy](../CONTRIBUTING.md#ai-assisted-contribution-policy) and check all boxes below before submitting your PR for review:**

- [x] I am a human, and not a bot.
- [x] I will be responsible for responding to review comments in a timely manner.
- [x] I will work with the maintainers to push this PR forward until submission.

**Note:** Failing to adhere to this agreement may result in your future PRs no longer being reviewed.

### CODE REVIEW

File: keras/src/backend/tensorflow/optimizer.py [hunk 1]
```diff
@@ -21,6 +21,10 @@ def __init__(self, *args, **kwargs):
         super().__init__(*args, **kwargs)
         self._distribution_strategy = tf.distribute.get_strategy()
 
+    def add_variable(self, *args, **kwargs):
+        with tf.init_scope():
+            return super().add_variable(*args, **kwargs)
```
[the remark this answers is not in the document]
Comment (peinguim) [reply keras/src/backend/tensorflow/optimizer.py]: Hmmm, overriding `build` as proposed doesn't solve the error from #23085.
But maybe there's a better place to enter `tf.init_scope()`?
I don't see for sure how it's incompatible with `colocate_vars_with` though 🤔

**Comment by peinguim** [comment]
About tests, if I add a `@tf.function` decorator to this existing one here it fails in master (with the same error as in #23085) and passes in the branch.

https://github.com/keras-team/keras/blob/76c1585f6a79d2a425b96a160f8bbc015beb6408/keras/src/backend/tensorflow/optimizer_distribute_test.py#L62-L64

If you think this deserves better test coverage, I could just add the decorator, or add a parameter to the test to enable it, or add a new dedicated test.

File: keras/src/backend/tensorflow/optimizer.py [hunk 2]
```diff
@@ -21,6 +21,10 @@ def __init__(self, *args, **kwargs):
         super().__init__(*args, **kwargs)
         self._distribution_strategy = tf.distribute.get_strategy()
 
+    def add_variable(self, *args, **kwargs):
```
Comment (hertschuh) [inline keras/src/backend/tensorflow/optimizer.py]: Can you copy the actual signature from the super class?

```python
    def add_variable(
        self,
        shape,
        initializer="zeros",
        dtype=None,
        aggregation="none",
        layout=None,
        name=None,
    ):
        ...
```

**Review by hertschuh — COMMENTED** [review]
Thanks a lot for figuring this out!

File: keras/src/backend/tensorflow/optimizer.py (hunk 2 above)
Comment (peinguim) [reply keras/src/backend/tensorflow/optimizer.py]: Thanks for the feedback!
Done in https://github.com/peinguim/keras/commit/85b1fc73a3623e8bbfa91e545522f6fb25ff2ab1

**Review by hertschuh — APPROVED** [review]
Thank you for the fix!
